haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,216 @@
+from typing import List, Dict, Union, Tuple
+
+from sentence_transformers import CrossEncoder
+
+from haystack.nodes import BaseRetriever, QuestionGenerator
+from haystack.schema import Document
+from haystack.nodes.base import BaseComponent
+
+
+class PseudoLabelGenerator(BaseComponent):
+    """
+    The PseudoLabelGenerator is a component that creates Generative Pseudo Labeling (GPL) training data for training
+    of dense retrievers.
+
+    GPL is an unsupervised domain adaptation method for training of dense retrievers. It is based on question generation
+    and pseudo labeling with powerful cross-encoders. To train a domain-adapted model, it needs access to unlabeled
+    target corpus usually via DocumentStore and a retriever to mine for negatives.
+
+    For more details see https://github.com/UKPLab/gpl
+
+    For example:
+
+    ```python
+    |   document_store = DocumentStore(...)
+    |   retriever = Retriever(...)
+    |   qg = QuestionGenerator(model_name_or_path="doc2query/msmarco-t5-base-v1")
+    |   plg = PseudoLabelGenerator(qg, retriever)
+    |   train_examples = []
+    |   for idx, doc in enumerate(document_store):
+    |   output_samples = plg.run(documents=[doc])
+    |   for item in output_samples:
+    |       train_examples.append(item)
+    |
+    |   # Users can also provide a list of questions/document dictionary pairs to generate pseudo labels for
+    |
+    |   question_doc_pairs: List[Dict[str, str]] = [{"question": "question text ...", "document": "document text ..."}]
+    |   plg = PseudoLabelGenerator(question_doc_pairs, retriever)
+    |   train_examples = []
+    |   for idx, doc in enumerate(document_store):
+    |   output_samples = plg.run(documents=[doc])
+    |   for item in output_samples:
+    |       train_examples.append(item)
+    ```
+    """
+
+    def __init__(
+        self,
+        question_producer: Union[QuestionGenerator, List[Dict[str, str]]],
+        retriever: BaseRetriever,
+        cross_encoder_model_name_or_path: str = "cross-encoder/ms-marco-MiniLM-L-6-v2",
+        total_number_of_questions: int = sys.maxsize,
+        top_k: int = 10,
+    ):
+        """
+        Loads the cross encoder model and prepares PseudoLabelGenerator.
+
+        :param question_producer: The question producer used to generate questions or a list of already produced
+        questions/document pairs in Dict format {"question": "question text ...", "document": "document text ..."}.
+        :type question_producer: Union[QuestionGenerator, List[Dict[str, str]]]
+        :param retriever: The retriever used to query document stores
+        :type retriever: BaseRetriever
+        :param cross_encoder_model_name_or_path: The path to the cross encoder model, defaults to
+        cross-encoder/ms-marco-MiniLM-L-6-v2
+        :type cross_encoder_model_name_or_path: str (optional)
+        :param total_number_of_questions: This is the total number of questions that will be generated, defaults
```

Comment:
Why `total_number_of_questions`? In the paper they use 3 questions per document so shouldn't this rather be `num_questions_per_document`? And the total number of questions will result from the number of documents passed in * `num_questions_per_document`?

Reply:
+1 good point. But what if documents are long and there are a few of them. Perhaps we need both...

Reply:
If documents are long we will need to split them anyways, as the model will only use the first 512 tokens.

Reply:
Ok, 3 questions per 512 tokenized segments sounds reasonable.  Yes, we can remove total_number_of_questions. Would you still call the parameter `num_questions_per_document` or something else?

Reply:
I think `num_questions_per_document` describes it quite well. What do you think?

Reply:
Adding `max_questions_per_document`  and then remove any questions exceeding that number on a per-document basis